src/sources/vector.rs

```diff
@@ -43,13 +41,8 @@ inventory::submit! {
 
 #[typetag::serde(name = "vector")]
 impl SourceConfig for VectorConfig {
-    fn build(
-        &self,
-        _name: &str,
-        _globals: &GlobalOptions,
-        shutdown: ShutdownSignal,
-        out: mpsc::Sender<Event>,
-    ) -> crate::Result<super::Source> {
+    fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        let SourceContext { shutdown, out, .. } = cx;
```

Comment:
Not sure if I'm alone here, but I don't find this early destructuring pattern to be useful. It's visually somewhat jarring and spatially very isolated from where the bindings are actually used. I don't see any reason to maintain this upfront list of fields we're going to use later in the method as opposed to simply referring to them where needed.

Reply:
The rationale for this pattern is we're essentially replacing the arguments list. Since in my view the context should be just a plain struct, essentially a typed envelope for the arguments list of a particular function (`build`).
I hope this explains the pattern I'm using here.
Now, are there any downsides to it? I don't see a problem with arguments being spatially isolated - we had the same exact behavior with just plan arguments.

The reason why I like the context to be just an envelope for the args is because it's the ultimate simplicity you could get here. It behaves the same way, but has a single additional benefit of the ability to pass the bundled data together downstream. It's not free, but the difference in generated asm is very negligable:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8304d0f4b5c0e7a7cf477d64695b2d34
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9e93c42c2ce66de643524a7372811992


The vase of my choice of this pattern here is I think we don't really want to even try to be smart about what context is, and just keep is simple.

Reply:
> The reason why I like the context to be just an envelope for the args is that it's the ultimate simplicity you could get here.

Wouldn't the simplest thing be to prefix each of these variables with `cx.` where they're used? I see no reason to try to avoid that. It seems like you've invented a pattern to avoid accessing struct fields and I just don't see why.

> When this pattern results in such growth of the LOCs it's clearly not a win.

Unless you're talking about additional linebreaks due to `cx.` making lines longer, this pattern fundamentally always introduces additional lines of code.

> in our real-world case it's actually better than just specifying the args cause we can omit imports

I don't understand this. There should be no difference in the required imports for this vs simple struct member access. If you're referring to adding these fields as arguments to the function instead of fields on the context then sure, but I don't think anyone is arguing for that.

Reply:
> Wouldn't the simplest thing be to prefix each of these variables with `cx.` where they're used? I see no reason to try to avoid that. It seems like you've invented a pattern to avoid accessing struct fields and I just don't see why.

There are cases where you get dereferencing-shared-object errors when using values directly off of `cx.`. Thus, destructuring explicitly takes ownership of all the values.

But in our case it doesn't seem like we'd have such issues, so, yeah, fair point.

The rest of the points are good too. The nvm about the loc, and the imports situation is the same as with the alternative cx designs.

Reply:
Regardless of whether or not we'd like to introduce this pattern or not, it's an additional unnecessary change that is distracting from the actual purpose of the PR. I'd be happy to continue discussing its pros and cons on a dedicated PR, but I think it's more important that we get `SourceContext` in place and move on.

Reply:
Agreed, we should split the concerns here. I assumed it wouldn't be so controversial, bad call. 😄